src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -261,27 +261,42 @@ static const Type* bitshuffle_value(const TypeInteger* src_type, const TypeInteg
   }
 
   if (!mask_type->is_con()) {
-    int mask_max_bw;
-    int max_bw = bt == T_INT ? 32 : 64;
-    // Case 1) Mask value range includes -1.
-    if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
-      mask_max_bw = max_bw;
-    // Case 2) Mask value range is less than -1.
-    } else if (mask_type->hi_as_long() < -1L) {
-      mask_max_bw = max_bw - 1;
-    } else {
-    // Case 3) Mask value range only includes +ve values.
-      assert(mask_type->lo_as_long() >= 0, "");
-      jlong clz = count_leading_zeros(mask_type->hi_as_long());
-      clz = bt == T_INT ? clz - 32 : clz;
-      mask_max_bw = max_bw - clz;
-    }
     if ( opc == Op_CompressBits) {
-      lo = mask_max_bw == max_bw ? lo : 0L;
-      // Compress operation is inherently an unsigned operation and
-      // result value range is primarily dependent on true count
-      // of participating mask value.
-      hi = mask_max_bw < max_bw ? (1L << mask_max_bw) - 1 : src_type->hi_as_long();
+      // Pattern: Integer/Long.compress(src_type, mask_type)
+      int max_mask_bit_width;
+      int mask_bit_width = bt == T_INT ? 32 : 64;
+      if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
+        // Case 1) Mask value range includes -1, this negates the possibility of
+        // strictly non-negative result value range.
+        max_mask_bit_width = mask_bit_width;
+      } else if (mask_type->hi_as_long() < -1L) {
+        // Case 2) Mask value range is less than -1, this indicates presence of at least
+        // one zero bit in the mask value, thereby constraining the result of compression
+        // to a +ve value range.
+        max_mask_bit_width = mask_bit_width - 1;
+      } else {
+        // Case 3) Mask value range only includes +ve values, thus we can
+        // identify leading known zero bits of mask value and use this to
+        // constrain upper bound of result value range.
+        assert(mask_type->lo_as_long() >= 0, "");
+        // Here, result of clz is w.r.t to long argument, hence for integer argument
+        // we explicitly subtract 32 from the result.
```

Comment:
The comments are not at the correct line. It does not seem to refer to the line above or below. I would say either move it up to the `Case 3` comments, or move it down to the specific line that actually subtracts `32`.